wgsl/index.bs

```diff
@@ -922,9 +945,151 @@ a result value, such as the non-result-value effects of its subexpressions.
 
 TODO: example: non-result-value effect is any side effect of a function call subexpression.
 
+## Ideal Types ## {#ideal-types-sec}
+
+There are two <dfn noexport>ideal types</dfn>:
+* The <dfn noexport>IdealInteger</dfn> type is the set of all integers.
+    * An [=syntax/int_literal=] without an `i` or `u` suffix denotes an [=IdealInteger=] value.
+* The <dfn noexport>IdealReal</dfn> type is the set of all real numbers (e.g. 1.25, -100, &pi;)
+    * A [=syntax/float_literal=] without an `f` suffix denotes a [=IdealReal=] value.
+
+A type is <dfn noexport>concrete</dfn> if it is not [=ideal type|ideal=].
+
+Ideal types support a limited set of operations: parenthesization, and unary minus.
+
+<table class='data'>
+  <caption>Expressions on ideal types</caption>
+  <thead>
+    <tr><th>Precondition<th>Conclusion<th>Description
+  </thead>
+  <tr algorithm="ideal negation">
+    <td>|e|: |T|<br>
+        |T| is an [=ideal type=]
+    <td>`-`|e|`:` |T|
+        <td>Negation: Evaluates to 0 - |e|.
+
+  <tr algorithm="ideal parenthesization">
+    <td>|e|: |T|<br>
+        |T| is an [=ideal type=]
+    <td>`(`|e|`)` : |T|
+    <td>Parenthesization: Evaluates to |e|.
+</table>
+TODO: Consider const-expr over ideal types.
```

Comment:
In offline feedback, @alan-baker wanted this to be far more generally expanded, covering operations (e.g. + -, etc), and builtin functions.  That seemed the direction desired by the group.